### DISCUSSION TITLE
How to toggle `copilot.nvim` and `copilot-cmp.nvim`?

### DISCUSSION THREAD

**Discussion by Saplyn** [LazyVim/LazyVim#4232] [Q&A] [answered]
I would like to disable the copilot completion ([in LazyExtra](https://www.lazyvim.org/extras/coding/copilot)) when I don't need it, and enable it back when I feel like to (toggle it). However, several attempt to do this myself failed.

I have tried:

- Adding a [key](https://lazy.folke.io/spec/lazy_loading#%EF%B8%8F-lazy-key-mappings) to toggle `zbirenbaum/copilot.lua`, but found that whenever I enter insert mode it pops up back again.
- Trys to disable/toggle `zbirenbaum/copilot-cmp`, but I'm too stupid to figure that out.
- The current workaround would be: `:LazyExtra`, disable/enable `coding.copilot`, `:q`, `nvim`. But that's a lot of steps.

I have read:

- [Toggle plugin](https://github.com/LazyVim/LazyVim/discussions/3767), but found no reply.
- [Start copilot default disabled](https://github.com/LazyVim/LazyVim/discussions/1151), but I can't get this to work (invalid lazy spec).

Is this possible outside of LazyExtra config? Also, should this functionality be added to Extra?

**Comment by jellis206** [comment]
The setup I have works for me, I even wrote an additional command to toggle autocomplete

[Link to my repo](https://github.com/jellis206/nvim)

Here is my keymap that I added using which key [/lua/config/keymaps.lua](https://github.com/jellis206/nvim/blob/main/lua/config/keymaps.lua)
```lua
  { "<leader>uc", "<cmd>ToggleCopilot<cr>", desc = "Toggle Copilot" },
```
Here is the command that I add in my [/plugins/auto-complete.lua](https://github.com/jellis206/nvim/blob/main/lua/plugins/auto-complete.lua) which I do in an options function, you could probably do this in a copilot setup, I have my own because I didn't like the autocomplete stuff the default way.
```lua
      vim.cmd("Copilot disable")
      local copilot_enabled = false
      vim.api.nvim_create_user_command("ToggleCopilot", function()
        if copilot_enabled then
          vim.cmd("Copilot disable")
          copilot_enabled = false
        else
          vim.cmd("Copilot enable")
          copilot_enabled = true
        end
      end, {})
```

**Reply by jellis206** [reply]
~~my setup starts with it enabled but you could probably use the vim.cmd somewhere to disable it on startup~~

**Reply by jellis206** [reply]
I actually just changed that by adding a line that disables copilot and it worked :)

**Reply by Saplyn** [reply]
Thanks a lot! I was about to post my config (~~I figured it out 10 minutes ago, then I went to take a shower and decided to post after that~~).

After some workaround I found that the `Copilot toggle` command seems to be broken (the copilot re-enables itself shortly afterward), but the `Copilot enable` and `Copilot disable` works as expected, and the `require("copilot.client").is_disabled()` could provide the current status of copilot, so:

```lua
-- lua/plugins/copilot.lua
return {
  "zbirenbaum/copilot.lua",
  keys = {
    {
      "<leader>at",
      function()
        if require("copilot.client").is_disabled() then
          require("copilot.command").enable()
        else
          require("copilot.command").disable()
        end
      end,
      desc = "Toggle (Copilot)",
    },
  },
}
```

This could toggle copilot regardless of it started enabled or disabled.

But still thanks for sharing!

**Comment by Saplyn** [comment]
For those who wants the `whick-key.nvim` hint to display the "copilot" icon, this is the updated config:

```lua
-- lua/plugins/copilot.lua
return {
  "zbirenbaum/copilot.lua",
  keys = {
    {
      "<leader>at",
      function()
        if require("copilot.client").is_disabled() then
          require("copilot.command").enable()
        else
          require("copilot.command").disable()
        end
      end,
      desc = "Toggle (Copilot)",
    },
  },
}
```

```lua
-- lua/plugins/which-key.lua
return {
  "folke/which-key.nvim",
  opts = {
    icons = {
      rules = {
        { plugin = "copilot.lua", icon = " ", color = "orange" },
      },
    },
  },
}
```

Or you may want to combine it into one, but I prefer one-file-per-plugin :)

**Comment by pogopaule** [comment]
```lua
{
  "zbirenbaum/copilot-cmp",
  opts = function()
    local copilot_toggle = require("lazyvim.util.toggle").wrap({
      name = "Copilot Completion",
      get = function()
        return not require("copilot.client").is_disabled()
      end,
      set = function(state)
        if state then
          require("copilot.command").enable()
        else
          require("copilot.command").disable()
        end
      end,
    })

    LazyVim.toggle.map("<leader>at", copilot_toggle)
  end,
},

```

![image](https://github.com/user-attachments/assets/c6160a07-977f-453f-ab6a-02e516541228)

**Reply by ficd0** [reply]
especially configuring with the lazyvim toggle is really nice man. I'm wondering, do you know if there's a way to disable the notification that happens when you toggle this option? Because it's also indicated by the little copilot icon in the bottom right corner of the lualine, I find the additional "copilot disabled!!!" notification pretty distracting. Is there some kind of "silent" option for the toggle or?

**Reply by dpetka2001** [reply]
@ficcdaf I created #4655 to accommodate this. If it gets accepted, you would just have to add `silent = true` to your toggle to disable the notification pop-up. 

PS: You can also use `icon` and `color_enabled`/`color_disabled` already to further customize it to your liking. See my [comment](https://github.com/LazyVim/LazyVim/discussions/4512#discussioncomment-10813289) here for an example.

**Reply by ficd0** [reply]
@dpetka2001 Thanks for this. I reviewed your PR, and I hope it gets accepted soon.

**Comment by ficd0** [comment]
thank you so much, not being able to toggle this is why I stopped using copilot about a year or so back before it was part of lazyvim, this is a life saver

**Comment by exsesx** [comment]
Here's the solution I came up with:

```lua
-- keymaps.lua or, in my case, keymaps/copilot.lua
local copilot_exists = pcall(require, "copilot")

if copilot_exists then
  LazyVim.toggle.map("<leader>at", {
    name = "Copilot Completion",
    color_enabled = "azure",
    color_disabled = "orange",
    get = function()
      return not require("copilot.client").is_disabled()
    end,
    set = function(state)
      if state then
        require("copilot.command").enable()
      else
        require("copilot.command").disable()
      end
    end,
  })
end
```

**Reply by exsesx** [reply, marked as answer]
Since [LazyVim 13.0](https://github.com/LazyVim/LazyVim/releases/tag/v13.0.0) and introduction of [snacks.nvim](https://github.com/folke/snacks.nvim):

```lua
local Snacks = require("snacks")
local copilot_exists = pcall(require, "copilot")

if copilot_exists then
  Snacks.toggle({
    name = "Copilot Completion",
    color = {
      enabled = "azure",
      disabled = "orange",
    },
    get = function()
      return not require("copilot.client").is_disabled()
    end,
    set = function(state)
      if state then
        require("copilot.command").enable()
      else
        require("copilot.command").disable()
      end
    end,
  }):map("<leader>at")
end
```

**Reply by folke** [reply]
Looks good! But color is an object though. You may want to change that.

**Reply by exsesx** [reply]
> Looks good! But color is an object though. You may want to change that.

Oh, you're right, thanks! Updated.

Thanks for all your hard work 🙏

**Reply by b11e** [reply]
For people who wonder where to put this in your lazyvim config, I went and added it into a `snacks.lua` file in `plugins/`

```lua
return {
  "folke/snacks.nvim",
  priority = 1000,
  lazy = false,
  ---@type snacks.Config
  opts = {
    bigfile = { enabled = true },
    dashboard = { enabled = true },
    explorer = { enabled = true },
    indent = { enabled = true },
    input = { enabled = true },
    notifier = {
      enabled = true,
      timeout = 4000,
    },
    picker = { enabled = true },
    quickfile = { enabled = true },
    scope = { enabled = true },
    scroll = { enabled = true },
    statuscolumn = { enabled = true },
    words = { enabled = true },
    styles = {
      notification = {
        wo = { wrap = true }, -- Wrap notifications
      },
    },
  },
  keys = {
    {
      "<leader>N",
      desc = "Neovim News",
      function()
        Snacks.win({
          file = vim.api.nvim_get_runtime_file("doc/news.txt", false)[1],
          width = 0.6,
          height = 0.6,
          wo = {
            spell = false,
            wrap = false,
            signcolumn = "yes",
            statuscolumn = " ",
            conceallevel = 3,
          },
        })
      end,
    },
  },
  init = function()
    vim.api.nvim_create_autocmd("User", {
      pattern = "VeryLazy",
      callback = function()
        -- Setup some globals for debugging (lazy-loaded)
        _G.dd = function(...)
          Snacks.debug.inspect(...)
        end
        _G.bt = function()
          Snacks.debug.backtrace()
        end
        vim.print = _G.dd -- Override print to use snacks for `:=` command

        -- Create some toggle mappings
        Snacks.toggle.option("spell", { name = "Spelling" }):map("<leader>us")
        Snacks.toggle.option("wrap", { name = "Wrap" }):map("<leader>uw")
        Snacks.toggle.option("relativenumber", { name = "Relative Number" }):map("<leader>uL")
        Snacks.toggle.diagnostics():map("<leader>ud")
        Snacks.toggle.line_number():map("<leader>ul")
        Snacks.toggle
          .option("conceallevel", { off = 0, on = vim.o.conceallevel > 0 and vim.o.conceallevel or 2 })
          :map("<leader>uc")
        Snacks.toggle.treesitter():map("<leader>uT")
        Snacks.toggle.option("background", { off = "light", on = "dark", name = "Dark Background" }):map("<leader>ub")
        Snacks.toggle.inlay_hints():map("<leader>uh")
        Snacks.toggle.indent():map("<leader>ug")
        Snacks.toggle.dim():map("<leader>uD")

        -- Add Copilot toggle
        local copilot_exists = pcall(require, "copilot")
        if copilot_exists then
          Snacks.toggle({
            name = "Copilot Completion",
            color = {
              enabled = "azure",
              disabled = "orange",
            },
            get = function()
              return not require("copilot.client").is_disabled()
            end,
            set = function(state)
              if state then
                require("copilot.command").enable()
              else
                require("copilot.command").disable()
              end
            end,
          }):map("<leader>at")
        end
        --
      end,
    })
  end,
}

```

**Reply by tionichm** [reply]
I'm using `copilot.lua`. 

I had to adapt the above code a little bit, but it ends up doing what we want.

```lua
--- ~/.config/nvim/lua/plugins/snacks.nvim
return {
  "folke/snacks.nvim",
  --- Updated this section to align with approach taken in other LazyVim snacks configs
  ---@diagnostic disable-next-line: unused-local
  opts = function(_, opts)
    -- Credit for this section goes to exsesx
    -- https://github.com/LazyVim/LazyVim/discussions/4232#discussioncomment-11191278
    local snacks = require("snacks")
    -- Check whether Copilot is installed
    if pcall(require, "copilot") then
      --- Workaround to keep track of state
      vim.g.snacks_copilot_enabled = true
      snacks
        .toggle({
          name = "Toggle (Copilot Completion)",
          color = {
            enabled = "azure",
            disabled = "orange",
          },
          get = function()
            return vim.g.snacks_copilot_enabled
          end,
          set = function(state)
            if state then
              vim.g.snacks_copilot_enabled = true
              require("copilot.command").enable()
            else
              vim.g.snacks_copilot_enabled = false
              require("copilot.command").disable()
            end
          end,
        })
        :map("<leader>at")
    end
  end,
}
```

**Reply by CleisonManriqueAguirre** [reply]
Thanks !!! The only setup to enable disable for those works installed from lazyextras
